In [23]:
import pandas as pd
import numpy as np
df=pd.read_csv("/content/drive/MyDrive/datasetsworkshop1/student.csv")


3.1 implementation from scratch step by step guide

In [24]:
#print the top 5 of the dataset
df.head()


Math  Reading  Writing
0    48       68       63
1    62       81       72
2    79       80       78
3    76       83       79
4    59       64       62

In [25]:
#print the bottom 5 of the dataset
df.tail()

Math  Reading  Writing
995    72       74       70
996    73       86       90
997    89       87       94
998    83       82       78
999    66       66       72

In [26]:
#print the info of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   Math     1000 non-null   int64
 1   Reading  1000 non-null   int64
 2   Writing  1000 non-null   int64
dtypes: int64(3)
memory usage: 23.6 KB


In [27]:
#gather descroptive info about the dataset
df.describe()

Math      Reading      Writing
count  1000.000000  1000.000000  1000.000000
mean     67.290000    69.872000    68.616000
std      15.085008    14.657027    15.241287
min      13.000000    19.000000    14.000000
25%      58.000000    60.750000    58.000000
50%      68.000000    70.000000    69.500000
75%      78.000000    81.000000    79.000000
max     100.000000   100.000000   100.000000

In [28]:
#split the data into features and labels
X=df.drop(columns=['Writing']).values
Y=df['Writing'].values


In [29]:
#creating the matrix
d = X.shape[1]
W = np.random.rand(d, 1)
Y_hat = np.dot(W.T, X.T).T

print("W:")
print(W)
print("X:")
print(X)
print("Y_hat:")
print(Y_hat)


W:
[[0.94530153]
 [0.86948853]]
X:
[[48 68]
 [62 81]
 [79 80]
 ...
 [89 87]
 [83 82]
 [66 66]]
Y_hat:
[[104.49969368]
 [129.03726605]
 [144.23790359]
 [144.01046458]
 [111.42005643]
 [138.26284238]
 [139.20814391]
 [ 85.21932001]
 [125.48349892]
 [149.11603726]
 [118.60340368]
 [144.5767011 ]
 [ 91.19438122]
 [144.38952959]
 [118.52759068]
 [ 84.57727048]
 [118.45177768]
 [ 84.27401847]
 [145.37037663]
 [136.1092548 ]
 [112.82023598]
 [111.26843042]
 [125.93837694]
 [104.7674002 ]
 [139.0565179 ]
 [142.53447203]
 [115.99493809]
 [118.14852567]
 [145.33483113]
 [131.79735765]
 [134.06702573]
 [143.47977356]
 [124.8414494 ]
 [131.83762515]
 [ 93.31242329]
 [ 97.46797244]
 [ 93.34796879]
 [137.84823186]
 [129.56795707]
 [ 77.96015975]
 [144.6525141 ]
 [140.68413646]
 [115.35288857]
 [ 90.8153162 ]
 [116.78861362]
 [127.82898001]
 [ 94.29327033]
 [124.8414494 ]
 [172.02599107]
 [119.1696402 ]
 [181.4790064 ]
 [146.39149116]
 [129.03726605]
 [ 86.88248406]
 [ 87.9035986 ]
 [ 92.70591927]
 [

In [30]:

def train_test_split(X, y, test_size=0.2, random_seed=42):
    """
    Splits dataset into train and test sets.
    Arguments:
    X : np.ndarray
        Feature matrix.
    y : np.ndarray
        Target array.
    test_size : float
        Proportion of the dataset to include in the test split (0 < test_size < 1).
    random_seed : int
        Seed for reproducibility.
    Returns:
    X_train, X_test, y_train, y_test : np.ndarray
        Training and testing splits of features and target.
    """
    np.random.seed(random_seed)
    indices = np.arange(X.shape[0])
    np.random.shuffle(indices)  # Shuffle the indices
    test_split_size = int(len(X) * test_size)
    test_indices = indices[:test_split_size]
    train_indices = indices[test_split_size:]
    X_train, X_test = X[train_indices], X[test_indices]
    y_train, y_test = y[train_indices], y[test_indices]
    return X_train, X_test, y_train, y_test

# Perform the train-test split with an 80-20 split
X_train, X_test, y_train, y_test = train_test_split(X, Y_hat, test_size=0.2)

# Output shapes to verify
print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)


Shape of X_train: (800, 2)
Shape of X_test: (200, 2)
Shape of y_train: (800, 1)
Shape of y_test: (200, 1)


In [31]:
#mean squared error

def cost_function(X, Y, W):
    y_pred = np.dot(X, W)
    squared_error = (y_pred - Y) ** 2
    cost = np.mean(squared_error) / 2
    return cost

X_test = np.array([[1, 2], [3, 4], [5, 6]])
Y_test = np.array([3, 7, 11])
W_test = np.array([1, 1])

cost = cost_function(X_test, Y_test, W_test)

print("Cost function output:", cost)

if np.isclose(cost, 0):
    print("no errors")
else:
    print("Something went wrong: Reimplement the cost function")


Cost function output: 0.0
no errors


In [32]:


def gradient_descent(X, Y, W, alpha, iterations):
    cost_history = [0] * iterations
    m = len(Y)

    for iteration in range(iterations):
        Y_pred = np.dot(X, W)
        loss = Y_pred - Y
        dw = (1/m) * np.dot(X.T, loss)
        W_update = W - alpha * dw
        cost = cost_function(X, Y, W_update)
        cost_history[iteration] = cost
        W = W_update

    return W_update, cost_history

np.random.seed(0)
X = np.random.rand(100, 3)
Y = np.random.rand(100)
W = np.random.rand(3)

alpha = 0.01
iterations = 1000

final_params, cost_history = gradient_descent(X, Y, W, alpha, iterations)

print("Final Parameters:", final_params)
print("Cost History:", cost_history)


Final Parameters: [0.20551667 0.54295081 0.10388027]
Cost History: [0.10711197094660153, 0.10634880599939901, 0.10559826315680616, 0.10486012948320558, 0.1041341956428534, 0.10342025583900626, 0.1027181077540776, 0.1020275524908062, 0.10134839451441931, 0.1006804415957737, 0.1000235047554587, 0.09937739820884377, 0.09874193931205609, 0.09811694850887098, 0.09750224927850094, 0.0968976680842672, 0.09630303432313951, 0.09571818027612913, 0.09514294105952065, 0.09457715457692842, 0.09402066147216397, 0.09347330508290015, 0.09293493139511913, 0.09240538899833017, 0.09188452904154543, 0.0913722051899995, 0.09086827358260123, 0.09037259279010502, 0.08988502377398917, 0.08940542984603007, 0.08893367662855953, 0.08846963201539432, 0.08801316613342668, 0.08756415130486386, 0.08712246201010665, 0.08668797485125507, 0.08626056851623205, 0.08584012374351278, 0.08542652328745133, 0.08501965188419301, 0.0846193962181636, 0.08422564488912489, 0.08383828837978763, 0.08345721902397185, 0.08308233097530

In [33]:
#to calculate the root mean squared error
import numpy as np

def rmse(Y, Y_pred):
    rmse = np.sqrt(np.mean((Y_pred - Y) ** 2))
    return rmse


In [34]:
import numpy as np

def r2(Y, Y_pred):
    mean_y = np.mean(Y)
    ss_tot = np.sum((Y - mean_y) ** 2)
    ss_res = np.sum((Y - Y_pred) ** 2)
    r2 = 1 - (ss_res / ss_tot)
    return r2


In [38]:
from sklearn.model_selection import train_test_split

def main():

    X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

    W = np.zeros(X_train.shape[1])
    alpha = 0.00001
    iterations = 1000

    W_optimal, cost_history = gradient_descent(X_train, Y_train, W, alpha, iterations)

    Y_pred = np.dot(X_test, W_optimal)

    model_rmse = rmse(Y_test, Y_pred)
    model_r2 = r2(Y_test, Y_pred)

    print("Final Weights:", W_optimal)
    print("Cost History (First 10 iterations):", cost_history[:10])
    print("RMSE on Test Set:", model_rmse)
    print("R-Squared on Test Set:", model_r2)

if __name__ == "__main__":
    main()


Final Weights: [0.00218645 0.00228096 0.0023401 ]
Cost History (First 10 iterations): [0.14050248691328254, 0.14050092808849488, 0.14049936928968468, 0.14049781051685148, 0.1404962517699949, 0.14049469304911452, 0.14049313435420985, 0.14049157568528048, 0.140490017042326, 0.1404884584253459]
RMSE on Test Set: 0.628436759461518
R-Squared on Test Set: -3.095766837337692


The model appears to be underfitting, as indicated by the negative R² value (-3.0957) and the high RMSE (0.6284). Moreover, the gradual decrease in the cost function suggests that the model is not effectively learning from the data.

The higher learining rate can speed up the learning process but if it is too high it may cause instability in the model.
a slow learing rate results in slow learinig and underfitting